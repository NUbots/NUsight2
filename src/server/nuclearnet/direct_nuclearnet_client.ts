import { NUClearNet } from 'nuclearnet.js'
import { NUClearNetOptions } from 'nuclearnet.js'
import { NUClearNetSend } from 'nuclearnet.js'
import { NUClearNetPeer } from 'nuclearnet.js'
import { NUClearPacketListener } from '../../shared/nuclearnet/nuclearnet_client'
import { NUClearEventListener } from '../../shared/nuclearnet/nuclearnet_client'
import { NUClearNetClient } from '../../shared/nuclearnet/nuclearnet_client'

/**
 * A thin adapter around the real NUClearNet which implements the NUClearNetClient interface.
 */
export class DirectNUClearNetClient implements NUClearNetClient {
  public constructor(private nuclearNetwork: NUClearNet) {
  }

  public static of(): DirectNUClearNetClient {
    const nuclearNetwork = new NUClearNet()
    return new DirectNUClearNetClient(nuclearNetwork)
  }

  public connect(options: NUClearNetOptions): () => void {
    this.nuclearNetwork.connect(options)
    return () => this.nuclearNetwork.disconnect()
  }

  public onJoin(cb: NUClearEventListener): () => void {
    this.nuclearNetwork.on('nuclear_join', cb)
    return () => this.nuclearNetwork.removeListener('nuclear_join', cb)
  }

  public onLeave(cb: NUClearEventListener): () => void {
    this.nuclearNetwork.on('nuclear_leave', cb)
    return () => this.nuclearNetwork.removeListener('nuclear_leave', cb)
  }

  public on(event: string, cb: NUClearPacketListener): () => void {
    this.nuclearNetwork.on(event, cb)
    return () => this.nuclearNetwork.removeListener(event, cb)
  }

  public onPacket(cb: NUClearPacketListener): () => void {
    return this.on('nuclear_packet', cb)
  }

  public send(options: NUClearNetSend): void {
    this.nuclearNetwork.send(options)
  }
}
