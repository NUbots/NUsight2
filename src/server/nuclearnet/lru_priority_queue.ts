type Item<K, V> = {
  key: K,
  values: V[]
}

export class LruPriorityQueue<K, V> {
  private readonly limitPerKey?: number
  private readonly queue: Array<Item<K, V>> = []
  private readonly map: Map<K, Item<K, V>> = new Map()

  constructor({ limitPerKey }: { limitPerKey?: number } = {}) {
    this.limitPerKey = limitPerKey
  }

  add(key: K, value: V) {
    let item = this.map.get(key)
    if (!item) {
      item = { key, values: [] }
      this.map.set(key, item)
      this.queue.push(item)
    }
    if (this.limitPerKey && item.values.length >= this.limitPerKey) {
      item.values.shift()
    }
    item.values.push(value)
  }

  pop(): V | undefined {
    /* tslint:disable-next-line prefer-for-of */
    for (let i = 0; i < this.queue.length; i++) {
      const item = this.queue.shift()!
      this.queue.push(item)
      const value = item.values.shift()
      if (value) {
        return value
      }
    }
  }
}
