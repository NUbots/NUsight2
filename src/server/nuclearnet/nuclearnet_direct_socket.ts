import { NUClearNet } from 'nuclearnet.js'
import { NUClearNetSocket } from '../../shared/nuclearnet/nuclearnet_types'
import { NUClearPacketListener } from '../../shared/nuclearnet/nuclearnet_types'
import { NUClearEventListener } from '../../shared/nuclearnet/nuclearnet_types'
import { NUClearNetOptions } from '../../shared/nuclearnet/nuclearnet_types'
import { NUClearNetSend } from '../../shared/nuclearnet/nuclearnet_types'

export class NUClearNetDirectSocket implements NUClearNetSocket {
  public constructor(private nuclearNetwork: NUClearNet) {
  }

  public static of(): NUClearNetDirectSocket {
    const nuclearNetwork = new NUClearNet()
    return new NUClearNetDirectSocket(nuclearNetwork)
  }

  public connect(options: NUClearNetOptions): () => void {
    this.nuclearNetwork.connect(options)
    return () => this.nuclearNetwork.disconnect()
  }

  public onJoin(cb: NUClearEventListener): () => void {
    this.nuclearNetwork.on('nuclear_join', cb)
    return () => this.nuclearNetwork.removeListener('nuclear_join', cb)
  }

  public onLeave(cb: NUClearEventListener): () => void {
    this.nuclearNetwork.on('nuclear_leave', cb)
    return () => this.nuclearNetwork.removeListener('nuclear_leave', cb)
  }

  public on(event: string, cb: NUClearPacketListener): () => void {
    this.nuclearNetwork.on(event, cb)
    return () => this.nuclearNetwork.removeListener(event, cb)
  }

  public send(options: NUClearNetSend): void {
    this.nuclearNetwork.send(options)
  }
}
