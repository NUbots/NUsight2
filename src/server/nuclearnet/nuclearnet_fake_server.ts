import * as EventEmitter from 'events'
import { NUClearNetPeer } from 'nuclearnet.js'
import { NUClearNetSend } from 'nuclearnet.js'
import { createSingletonFactory } from '../../shared/base/create_singleton_factory'

export class NUClearNetFakeServer extends EventEmitter {
  private peers: NUClearNetPeer[]

  public constructor() {
    super()

    this.peers = []
  }

  public static of = createSingletonFactory(() => {
    return new NUClearNetFakeServer()
  })

  public connect(peer: NUClearNetPeer): void {
    this.emit('nuclear_join', peer)
    this.peers.push(peer)
  }

  public disconnect(peer: NUClearNetPeer) {
    this.emit('nuclear_leave', peer)
    this.peers.splice(this.peers.indexOf(peer), 1)
  }

  public send(peer: NUClearNetPeer, opts: NUClearNetSend) {
    if (typeof opts.type === 'string') {
      const packet = {
        peer,
        payload: opts.payload,
      }
      // TODO (Annable): Support opts.target
      this.emit(opts.type, packet)
    }
  }
}
