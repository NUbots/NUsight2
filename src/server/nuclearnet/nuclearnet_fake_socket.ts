import { NUClearNetOptions } from 'nuclearnet.js'
import { NUClearNetSend } from 'nuclearnet.js'
import { NUClearNetSocket } from '../../shared/nuclearnet/nuclearnet_types'
import { NUClearEventListener } from '../../shared/nuclearnet/nuclearnet_types'
import { NUClearPacketListener } from '../../shared/nuclearnet/nuclearnet_types'
import { NUClearNetFakeServer } from './nuclearnet_fake_server'

export class NUClearNetFakeSocket implements NUClearNetSocket {
  public constructor(private server: NUClearNetFakeServer) {
  }

  public static of(): NUClearNetFakeSocket {
    return new NUClearNetFakeSocket(NUClearNetFakeServer.of())
  }

  public connect(options: NUClearNetOptions): () => void {
    throw new Error('Not implemented')
  }

  public onJoin(cb: NUClearEventListener): () => void {
    throw new Error('Not implemented')
  }

  public onLeave(cb: NUClearEventListener): () => void {
    throw new Error('Not implemented')
  }

  public on(event: string, cb: NUClearPacketListener): () => void {
    throw new Error('Not implemented')
  }

  public send(options: NUClearNetSend): void {
    throw new Error('Not implemented')
  }
}
