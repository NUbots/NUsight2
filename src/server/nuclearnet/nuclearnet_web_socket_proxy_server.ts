import { NUClearNetOptions } from 'nuclearnet.js'
import * as SocketIO from 'socket.io'
import { NUClearNetClient } from '../../shared/nuclearnet/nuclearnet_types'
import { NUClearNetDirectSocket } from './nuclearnet_direct_socket'
import { NUClearNetFakeSocket } from './nuclearnet_fake_socket'
import SocketIOServer = SocketIO.Server
import SocketIOSocket = SocketIO.Socket
import Namespace = SocketIO.Namespace

type Options = {
  fakeNetworking: boolean
}

export class NUClearNetWebSocketProxyServer {
public constructor(private server: Namespace, private nuclearnetClient: NUClearNetClient) {
    server.on('connection', this.onClientConnection)
  }

  public static of(server: Namespace, opts: Options): NUClearNetWebSocketProxyServer {
    const nuclearnetClient: NUClearNetClient = opts.fakeNetworking ? NUClearNetFakeSocket.of() : NUClearNetDirectSocket.of()
    return new NUClearNetWebSocketProxyServer(server, nuclearnetClient)
  }

  private onClientConnection = (socket: SocketIOSocket) => {
    const listeners = new Map()

    const offJoin = this.nuclearnetClient.onJoin(peer => socket.emit('nuclear_join', peer))

    const offLeave = this.nuclearnetClient.onLeave(peer => socket.emit('nuclear_leave', peer))

    socket.on('nuclear_connect', (options: NUClearNetOptions) => {
      const disconnect = this.nuclearnetClient.connect(options)
      socket.on('nuclear_disconnect', () => disconnect())
    })

    socket.on('listen', (event: string, messageId: string) => {
      const off = this.nuclearnetClient.on(event, packet => socket.emit('nuclear_message', event, packet))

      let listenerIds = listeners.get(event)
      if (!listenerIds) {
        listenerIds = new Map()
        listeners.set(event, listenerIds)
      }
      listenerIds.set(messageId, off)
    })

    socket.on('unlisten', (event: string, messageId: string) => {
      const listenerIds = listeners.get(event)
      if (listenerIds) {
        const off = listenerIds.get(messageId)
        if (off) {
          off()
        }
      }
    })

    socket.on('disconnect', () => {
      offJoin()
      offLeave()
      for (const listenerIds of listeners) {
        for (const off of listenerIds.values()) {
          off()
        }
      }
    })
  }
}
