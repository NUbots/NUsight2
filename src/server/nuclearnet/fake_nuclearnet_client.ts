import { NUClearNetOptions } from 'nuclearnet.js'
import { NUClearNetSend } from 'nuclearnet.js'
import { NUClearEventListener } from '../../shared/nuclearnet/nuclearnet_client'
import { NUClearPacketListener } from '../../shared/nuclearnet/nuclearnet_client'
import { NUClearNetClient } from '../../shared/nuclearnet/nuclearnet_client'
import { FakeNUClearNetServer } from './fake_nuclearnet_server'
import { NUClearNetPeer } from 'nuclearnet.js'

export class FakeNUClearNetClient implements NUClearNetClient {
  private peer: NUClearNetPeer

  public constructor(private server: FakeNUClearNetServer) {
  }

  public static of(): FakeNUClearNetClient {
    return new FakeNUClearNetClient(FakeNUClearNetServer.of())
  }

  public connect(options: NUClearNetOptions): () => void {
    this.peer = {
      name: options.name,
      address: 'localhost',
      port: 1234,
    }
    this.server.connect(this.peer)

    return () => {
      this.server.disconnect(this.peer)
    }
  }

  public onJoin(cb: NUClearEventListener): () => void {
    return this.server.onJoin(cb)
  }

  public onLeave(cb: NUClearEventListener): () => void {
    return this.server.onLeave(cb)
  }

  public on(event: string, cb: NUClearPacketListener): () => void {
    throw new Error('Not implemented')
  }

  public send(options: NUClearNetSend): void {
    throw new Error('Not implemented')
  }
}
