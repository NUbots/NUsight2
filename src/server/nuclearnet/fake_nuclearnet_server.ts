import EventEmitter from 'events'
import { NUClearNetSend } from 'nuclearnet.js'
import XXH from 'xxhashjs'

import { createSingletonFactory } from '../../shared/base/create_singleton_factory'

import { FakeNUClearNetClient } from './fake_nuclearnet_client'

/**
 * A fake in-memory NUClearNet 'server' which routes messages between each FakeNUClearNetClient.
 *
 * All messages are 'reliable' in that nothing is intentially dropped.
 * Targeted messages are supported.
 */
export class FakeNUClearNetServer {
  private events: EventEmitter
  private clients: FakeNUClearNetClient[]

  constructor() {
    this.events = new EventEmitter()
    this.clients = []
  }

  /**
   * Designed to only have a single instance during runtime. This singleton is mimicking the behaviour of
   * the real NUClearNet, where a single global TCP/IP network is an implicit non-injectable dependency. This is why
   * FakeNUClearNetClient.of() does not take a given server and instead uses the singleton instance.
   *
   * Avoid using this singleton factory in tests though, as you'll introduce cross-contamination between tests.
   * Simply use the constructor of both FakeNUClearNetServer and FakeNUClearNetClient instead.
   */
  static of = createSingletonFactory(() => {
    return new FakeNUClearNetServer()
  })

  connect(client: FakeNUClearNetClient): () => void {
    this.events.emit('nuclear_join', client.peer)
    this.clients.push(client)

    for (const otherClient of this.clients) {
      // This ensures that the newly connected client receives a nuclear_join event from everyone else on the network.
      otherClient.peer && client.fakeJoin(otherClient.peer)

      // This conditional avoids sending nuclear_join twice to the newly connected client.
      if (otherClient !== client) {
        // Send a single nuclear_join to everyone on the network about the newly connected client.
        client.peer && otherClient.fakeJoin(client.peer)
      }
    }

    return () => {
      this.events.emit('nuclear_leave', client.peer)
      this.clients.splice(this.clients.indexOf(client), 1)

      for (const otherClient of this.clients) {
        client.peer && otherClient.fakeLeave(client.peer)
      }
    }
  }

  send(client: FakeNUClearNetClient, opts: NUClearNetSend) {
    if (!client.peer) {
      throw new Error('Cannot send a packaet to a client who has not connected to the network')
    }

    const hash: Buffer = typeof opts.type === 'string' ? hashType(opts.type) : opts.type
    const packet = {
      peer: client.peer,
      hash,
      payload: opts.payload,
      reliable: !!opts.reliable,
    }

    /*
     * This list intentionally includes the sender unless explicitly targeting another peer. This matches the real
     * NUClearNet behaviour.
     */
    const targetClients = opts.target === undefined
      ? this.clients
      : this.clients.filter(otherClient => otherClient.peer && otherClient.peer.name === opts.target)

    const hashString = hash.toString('hex')
    for (const client of targetClients) {
      client.fakePacket(hashString, packet)
    }
  }
}

export function hashType(type: string): Buffer {
  // Matches hashing implementation from NUClearNet
  // See https://goo.gl/6NDPo2
  let hashString: string = XXH.h64(type, 0x4e55436c).toString(16)
  // The hash string may truncate if it's smaller than 16 characters so we pad it with 0s
  hashString = ('0'.repeat(16) + hashString).slice(-16)

  return Buffer.from((hashString.match(/../g) as string[]).reverse().join(''), 'hex')
}
