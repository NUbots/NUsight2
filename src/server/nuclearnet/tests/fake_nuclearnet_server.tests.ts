import { FakeNUClearNetServer } from '../fake_nuclearnet_server'
import { createMockInstance } from '../../../shared/base/testing/create_mock_instance'
import { FakeNUClearNetClient } from '../fake_nuclearnet_client'
import Mocked = jest.Mocked
import MockInstance = jest.MockInstance
import { NUClearNetPeer } from 'nuclearnet.js'

describe.skip('FakeNUClearNetServer', () => {
  let server: FakeNUClearNetServer

  beforeEach(() => {
    server = new FakeNUClearNetServer()
  })

  it('sends nuclear_join events to new peers for all existing peers', () => {
    // TODO (Annable): Make FakeNUClearNetClient reasonable to mock and avoid the casting.
    const bob: Mocked<FakeNUClearNetClient> = createMockInstance(FakeNUClearNetClient)
    bob.peer = {
      name: 'bob',
      address: '127.0.0.1',
      port: 7447,
    } as any as NUClearNetPeer & MockInstance<NUClearNetPeer>
    server.connect(bob)

    const alice: Mocked<FakeNUClearNetClient> = createMockInstance(FakeNUClearNetClient)
    alice.peer = {
      name: 'alice',
      address: '127.0.0.1',
      port: 7447,
    } as any as NUClearNetPeer & MockInstance<NUClearNetPeer>
    server.connect(alice)

    const eve: Mocked<FakeNUClearNetClient> = createMockInstance(FakeNUClearNetClient)
    eve.peer = {
      name: 'eve',
      address: '127.0.0.1',
      port: 7447,
    } as any as NUClearNetPeer & MockInstance<NUClearNetPeer>
    server.connect(eve)

    expect(bob.fakeJoin).toHaveBeenCalledTimes(1)
    expect(bob.fakeJoin).toHaveBeenCalledWith(expect.objectContaining({ name: 'bob' }))

    expect(alice.fakeJoin).toHaveBeenCalledTimes(2)
    expect(alice.fakeJoin).toHaveBeenCalledWith(expect.objectContaining({ name: 'bob' }))
    expect(alice.fakeJoin).toHaveBeenCalledWith(expect.objectContaining({ name: 'alice' }))

    expect(eve.fakeJoin).toHaveBeenCalledTimes(3)
    expect(eve.fakeJoin).toHaveBeenCalledWith(expect.objectContaining({ name: 'bob' }))
    expect(eve.fakeJoin).toHaveBeenCalledWith(expect.objectContaining({ name: 'alice' }))
    expect(eve.fakeJoin).toHaveBeenCalledWith(expect.objectContaining({ name: 'eve' }))
  })
})
