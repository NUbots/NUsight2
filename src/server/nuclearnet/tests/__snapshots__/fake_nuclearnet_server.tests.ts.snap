// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FakeNUClearNetServer #hashType matches previous snapshots 1`] = `
Object {
  "data": Array [
    225,
    233,
    240,
    189,
    213,
    194,
    172,
    79,
  ],
  "type": "Buffer",
}
`;

exports[`FakeNUClearNetServer #hashType matches previous snapshots 2`] = `
Object {
  "data": Array [
    188,
    214,
    134,
    145,
    61,
    19,
    178,
    202,
  ],
  "type": "Buffer",
}
`;

exports[`FakeNUClearNetServer #hashType matches previous snapshots 3`] = `
Object {
  "data": Array [
    11,
    128,
    31,
    158,
    81,
    99,
    174,
    59,
  ],
  "type": "Buffer",
}
`;

exports[`FakeNUClearNetServer #hashType matches previous snapshots 4`] = `
Object {
  "data": Array [
    89,
    255,
    223,
    146,
    90,
    203,
    129,
    10,
  ],
  "type": "Buffer",
}
`;
